pyo3-macros-backend/src/pyclass.rs

```diff
@@ -1637,6 +1676,51 @@ fn impl_pytypeinfo(
     }
 }
 
+fn pyclass_richcmp(
+    options: &PyClassPyO3Options,
+    cls: &syn::Type,
+    ctx: &Ctx,
+) -> (Option<syn::ImplItemFn>, Option<MethodAndSlotDef>) {
+    let Ctx { pyo3_path } = ctx;
+
+    let eq_arms = options
+        .eq
+        .map(|eq| {
+            quote_spanned! { eq.span() =>
+                #pyo3_path::pyclass::CompareOp::Eq => {
+                    ::std::result::Result::Ok(#pyo3_path::conversion::IntoPy::into_py(self == other, py))
+                },
+                #pyo3_path::pyclass::CompareOp::Ne => {
+                    ::std::result::Result::Ok(#pyo3_path::conversion::IntoPy::into_py(self != other, py))
+                 },
+            }
+        })
+        .unwrap_or_default();
+
+    // TODO: `ord` can be integrated here (#4202)
+    if options.eq.is_some() {
+        let mut richcmp_impl = parse_quote! {
+            fn __pyo3__richcmp__(
```

Comment:
A thought - if we get this name to match up with what we generate from normal `__pymethods__` (I think it might be `__pymethod___richcmp____`), then we would be able to prevent the user from setting `pyclass(eq)` _and_ trying to implement `__richcmp__` themselves.

Reply:
I'll check that. I think it would be really nice if that errors.

Reply:
The error message is not super nice, but this works!

Reply:
It seems like the test for this confuses the type system such that one of our gil-refs deprecation warning is emitted. Any ideas how we should gate that test?

Reply:
Hmm confusing, that doesn't repro for me locally. Does it repro for you?

Reply:
Ah, I worked it out. So the cause was that in the `#[pymethods]` case the generated symbol taking raw pointers was `__pymethod_richcmp__`, which calls the user-provided `__richcmp__`.

In the macro we were emitting `__pymethod_richcmp__` as the equivalent of the `__richcmp__`, and then the function taking raw pointers (and doing the gil-refs checking) was called `__pymethod__default____pymethod_richcmp__` or something like that.

The problem is that these two `__pymethod_richcmp__` had different signatures which I think led to the type system getting confused and spitting out the warning.

So I pushed a commit which changes the names around so that `__pymethod_richcmp__` always takes the raw pointers (and do the gil-refs checking) and then the actual implementations are the user-provided function `__richcmp__` and PyO3 generated function `__pyo3_generated____richcmp__`.

While we still emit two `__pymethod_richcmp__` there's now no ambiguity in what they are trying to pass their arguments to, so type system no longer confused at the deprecation checking stage.

Reply:
Ah, I see, thanks for the fix 👍